Implementation of phrasal title casing (#326)

nschneid opened the pull request:
Per #326, protect.py now uses truelist-phrases to apply fixed-case to words that are sensitive to phrasal context. The new fixed-case logic is spelled out in the README.

Some technical notes:

- The `tokenize()` method now returns hyphen tokens rather than omitting them.
- A hyphen no longer counts as a separator that will trigger rule 5 to fixed-case the first word of the title.
- The code in protect.py that applies the fixed-case policy to the XML had to be rewritten. Previously it was matching based on word types, which led to false positives if a word appeared both in a truelisted phrase and outside the phrase. Now, a recased version of the original string is constructed and consulted to determine where to apply fixed-case in the XML.

Finally, the current truelist-phrases is fully manual; ideally we would mine abstracts to get a more comprehensive list of phrases that are regularly capitalized.

mjpost commented:
Hey, great! I’ll take a look soon. 

You can fix the build error by typing `make autofix` and committing the changes (we’re using [black](https://github.com/psf/black) and it often breaks builds).

mjpost commented:
One quick comment—can we merge `truelist` and `truelist-phrases`? I see no reason to distinguish them.

nschneid commented:
> One quick comment—can we merge `truelist` and `truelist-phrases`? I see no reason to distinguish them.

Done

nschneid commented:
OK, we have an issue with tokenization: NLTK's TreebankWordTokenizer is being used, but this escapes special characters like quotes, making it hard to align back to the XML. Would there be any harm in switching to another tokenizer (like Punkt)?

nschneid commented:
^ Fixed the tokenization issue with the `TreebankWordDetokenizer` + adding/removing spaces to the recased string until it matches the original ignoring case.

nschneid commented:
Added train_phrasal.py to mine abstracts for frequently capitalized phrases that appear in at least one title. I've added entries that should obviously be fixed-case to the truelist (43da612). Other suggestions output by the script are somewhat borderline and consist mostly of shared task names. Do we want to include those? Open to opinions about the items in 68923d6.

mjpost commented:
I am hoping @davidweichiang will have time to have a look, since he headed up the truecasing work. I'll get to this later this week.

nschneid commented:
I am working on improvements to the handling of all-caps. In the current XML data, stylistic all-caps sequences (excluding sequences of words that should be all-caps, such as acronyms) always seems to occur at the beginning or end of the title, and separated from the rest of the title by a punctuation token.

A heuristic that extracts such all-caps sequences, requiring at least 3 actual all-caps words (i.e. not punctuation) in the sequence, seems to work well.

Then there is the question of how the protection heuristics should work in a stylistic all-caps sequence. Currently protect.py is very conservative, basically not protecting any all-caps words if the title is majority all-caps. So I am thinking a better approach may be to follow the truelist capitalization for words that appear there (including, e.g., person and place names) and enforce all-caps for remaining words that do not appear in the falselist (general dictionary of English all-lowercase words), such as system names. This will be a problem for the few French titles that are in the Anthology, but c'est la vie?

@davidweichiang There is also protect_allcaps.py which uses the truelist and recognizes a few acronyms. How was this script applied—to all titles, but after protect.py, so in practice only affecting all-caps words that were not protected by protect.py?

nschneid commented:
I am working on improvements to the handling of all-caps. In the current XML data, stylistic all-caps sequences (excluding sequences of words that should be all-caps, such as acronyms) always seems to occur at the beginning or end of the title, and separated from the rest of the title by a punctuation token.

A heuristic that extracts such all-caps sequences, requiring at least 3 actual all-caps words (i.e. not punctuation) in the sequence, seems to work well.

Then there is the question of how the protection heuristics should work in a stylistic all-caps sequence. Currently protect.py is very conservative, basically not protecting any all-caps words if the title is majority all-caps. So I am thinking a better approach may be to follow the truelist capitalization for words that appear there (including, e.g., person and place names) and enforce all-caps for remaining words that do not appear in the falselist (general dictionary of English all-lowercase words), such as system names. This will be a problem for the few French titles that are in the Anthology, but c'est la vie?

@davidweichiang There is also protect_allcaps.py which uses the truelist and recognizes a few acronyms. How was this script applied—to all titles, but after protect.py, so in practice only affecting all-caps words that were not protected by protect.py?

nschneid commented:
> Still, the heuristic would fail on titles like "MORNING PANEL: Will NLP be solved in 10 years?" and "LSTM CRF POS tagging for Chinese".

Looking for a punctuation separator rules out the second one. The first one is indeed a problem due to the minimum-3-allcaps-words requirement. I'll have to see how many like that there are.

nschneid commented:
@davidweichiang Is "MORNING PANEL" an actual example? I can't find it in the XML files. The only ones I could find along those lines (2-word uses of stylistic all-caps) were:

- GRICE INCORPORATED
- INVITED TALK
- PROGRESS REPORT
- SINICA CORPUS

So we can special-case those in the script.

davidweichiang commented:
No, I made up both of those examples.

nschneid commented:
I'm printing lists of all-caps sequences flagged by the old and new heuristics. The true positive list is actually quite small, on the order of 50. I suggest we maintain a list of the all-caps titles and run the heuristics when new data is ingested to look for more.

nschneid commented:
There is also the question of what to do with single-character words that are capitalized in the raw title.

Abbreviations spelled with a period ("A.") are pretty much always personal names (esp. in book review titles and obituaries). There is also "U. S.". So these should always be protected.

Single-character capitals other than "A" and "I" include "N - gram", "K - best", "F - score", "Verb - Noun (V - N)", "E - Mail", and other miscellaneous uses in terms of art and hyphenated system names. Should we simply protect all of these single-character capitals? For some technical terms it is optional—authors may spell it "N-gram" or "n-gram" in the main text—but there are clearly cases where it is wrong not to capitalize, and I don't know of any that are clearly wrong to capitalize. (Capitalizing "E-mail" may look strange today, but it does occur like that in the main text of [this 2003 paper](https://www.aclweb.org/anthology/E03-1059.pdf), for example.)

nschneid commented:
(BTW I also checked for instances of multiletter capitalized words ending in a period according to the tokenizer. It's mostly "U.S.", so no need for a special rule.)

nschneid commented:
Nearly happy with the current version. However, I am not sure about the rule:

>  If a title starts with a single word set off by a colon or dash, the
>   first word is fixed-case unless the word appears in an English lexicon as
>   an all-lowercase lemma. (In principle this applies to both lowercase words
>   and uppercase words, but in practice protect.py never protects lowercase
>   characters, even title-initially.)

In "context2vec: Learning Generic Context Embedding with Bidirectional LSTM", should the lowercase "c" at the beginning be protected? What about the uppercase "E" for "Elissa: A Dialectal to Standard Arabic Machine Translation System"?

If we assume stylesheets will always print the first character of the title as is, we can remove this rule and the falselist that goes with it (which is incomplete anyway because it contains only lemmas).

nschneid commented:
@davidweichiang ^ Thoughts on whether title-initial characters should ever be fixed-case?

davidweichiang commented:
Yes, when a system name is written in all lowercase it may be correct to fixed-case it in my opinion. But it seems to me there are cases where the system name is capitalized when sentence initial. I am not sure if it should be up to the author or editor. 

> On Apr 29, 2020, at 01:20, Nathan Schneider <notifications@github.com> wrote:
> 
> ﻿
> @davidweichiang ^ Thoughts on whether title-initial characters should ever be fixed-case?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

nschneid commented:
OK, if nobody has a strong opinion on this I think I'll go with the more conservative option which is to not have a special rule for title-initial words, and leave it up to the stylesheet. That also simplifies the rules/code.

nschneid commented:
TODO: titles where parts of all-caps words are being protected after part was already fixed-case. Should rule 4 not apply if part of the word is already fixed-case?

```diff
-      <title>The <fixed-case>F</fixed-case>INITE <fixed-case>S</fixed-case>TRING Newsletter: Announcements</title>
+      <title>The <fixed-case>F</fixed-case><fixed-case>INITE</fixed-case> <fixed-case>S</fixed-case><fixed-case>TRING</fixed-case> Newsletter: Announcements</title>
```

nschneid commented:
Some stats on the changes:

2171 XML lines affected

```
   471 titles with "Twitter"
 + 131 with "Universal Dependen"
 +  73 with "Tree Adjoining Grammar"
 +  37 with "Combinatory Categorial Grammar"
 +  73 with "University"
 + 128 with non-word-initial capital (name or stylistic all-caps)
 + 414 with single uppercase character other than "A"
= 1327 (61%)
```

Most of the rest are due to truelist matches.

Once I fix the TODO above I should be done.

davidweichiang commented:
I could never figure out whether Finite String was supposed to be spelled in all caps or not.

nschneid commented:
@mjpost I think I'm happy with this implementation. When the code is merged I can submit a separate pull request for the data changes. 2176 lines of XML are affected (and most titles are on a single line).

mjpost commented:
Thanks, @nschneid, I have some other things looming but hope to get to this soon!

davidweichiang commented on bin/fixedcase/special-case-titles:
```diff
@@ -0,0 +1,98 @@
+the Finite String, volume 15, number 4
+the Finite String, volume 15, number 5
+the Finite String, volume 15, number 5 (continued)
+three titles from the Cambridge series: studies in natural language processing
```
Shouldn't the title of the series be capitalized?

nschneid replied:
"Studies in Natural Language Processing"? Yeah I guess

davidweichiang commented on bin/fixedcase/special-case-titles:
```diff
@@ -0,0 +1,98 @@
+ECONOMY IS A PERSON: a Chinese-English corpora and ontological-based comparison using the Conceptual Mapping Model
```
I didn't look at the paper, but are the first four words really always all caps?

nschneid replied:
That's part of the standard notation when naming a metaphor.

davidweichiang commented on bin/fixedcase/README.md:
```diff
@@ -1,27 +1,49 @@
-- If a title is in all caps, the above rule does not apply.
-- If a title starts with a single word set off by a colon or dash, the
-  first word is fixed-case.
+1. If a full title has a case-insensitive match in `special-case-titles`, only
```
I've always imagined the caser as being used once and then hand corrected. If the caser has a list like this, that means it's meant to be safe to run twice, right? does that need to be documented anywhere?

nschneid replied:
You mean someone might want to run it, then add entries to special-case-titles, then run it again?

davidweichiang replied:
I'll move this to the main comment thread...

davidweichiang commented:
Previously the imagined workflow would have been:

1. Run the caser on a newly ingested meeting
2. Correct any errors directly in the XML file
3. If there's another error, go to 2
4. But never run the caser again because it's not idempotent

And now I guess it's:

1. Run the caser on a newly ingested meeting
2. Correct any errors by modifying the special-case-titles list
3. Re-run the caser
4. If there's another error, go to 2
5. But never manually edit curly braces in this title again

It's my fault that the current workflow isn't well documented, but it'd probably be good to document the new workflow, especially 5, if I've understood correctly.

nschneid commented:
Oh I see what you mean. Actually, the caser never *removes* fixed-case, so it's always safe to add it manually to a particular title with an ambiguous word/phrase. The need for special-case-titles only really comes from stylistic all-caps (since the heuristics will assume any all-caps word other than "A" should be protected if not in special-case-titles).

So yes, it is safe to rerun provided that any titles with stylistic all-caps are added to special-case-titles. In fact, it's probably a good idea to rerun on the full anthology whenever the truelist is expanded with high-precision items like names of places/languages that may have been missed previously.

Suggested workflow:

1. Run the caser on a newly ingested meeting
2. Correct any errors as follows:
    - False Negatives: If a word or phrase should always have fixed-case capitals, add it to truelist. Otherwise, manually add `<fixed-case>` in the XML.
    - False Positives: If the heuristics for fixed-case produce a false positive for a title, add it to special-case-titles (lowercasing all but the fixed-case capitals). Simply removing `<fixed-case>` from the XML runs the risk that the change will be overridden on a subsequent run.
3. Re-run the caser on the full anthology if truelist has been modified, or just the newly ingested meeting otherwise. Repeat from step 2 if further errors are found.

davidweichiang commented:
That sounds good; where should it go for posterity?

mjpost commented:
Nathan, does this get called automatically in the current workflow? `ingest.py` is the script that gets used. It calls `normalize` (in `normalize_anth.py`), which calls `protect()` (from `fixedcase/protect.py`) on every "title" and "booktitle" field.

I think we could document this with a wiki page for now.

nschneid commented:
I've only run protect.py directly but haven't changed the signature of `protect()`, so I think it should work properly when called by the ingestion code.

Updated the README to elaborate on the workflow. Is that OK for now?

mjpost commented:
I tested this on some ACL files and it's really good. Thanks for your contribution! I hope you don't mind, I've added you to the volunteers list. I also alphabetized `truelist`.

mjpost commented:
I don't know why the diff says there are 412 files that differ. The actual diff against `master` is much smaller. Maybe it is a bug, and is diffing against *your* master branch?

Anyway, I'll merge this when the check completes.

nschneid commented:
I updated my master branch but that didn't fix it. Should I merge all the recent commits from master into nschneid:patch-8?

nschneid commented:
Oh, looks like you merged into a new branch phrase-casing. But I hadn't merged all the recent commits from master to nschneid:patch-8. Let me do that now.

mjpost commented:
I think it's okay. Do you want to merge? If so please do a "Squash and merge" and prune down the commit message to the bare essentials.

nschneid commented:
https://github.com/acl-org/acl-anthology/compare/master...nschneid:patch-8 shows the correct changes

nschneid commented:
Squash-and-merge seems to have worked! Showing 17 changed files.
